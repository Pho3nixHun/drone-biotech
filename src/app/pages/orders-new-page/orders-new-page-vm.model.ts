import { WithTitle } from '@interfaces/with-title.interface';
import { InputTextXVM } from '@components/input-text/input-text.component';
import { AreasDataFormControlVM } from './components/areas-data-form-control/areas-data-form-control.model';
import { ButtonXVM } from '@components/button/button.model';

interface AreasDataFormControlXVM extends AreasDataFormControlVM {
    addButtonXVM: ButtonXVM;
}

export interface OrdersNewPageVM extends WithTitle {
    areasDataFormControlXVM: AreasDataFormControlXVM;
    requiredAssistiveTextKey: string;
    internalOrderNumberInputTextXVM: InputTextXVM;
    nameInputTextXVM: InputTextXVM;
    phoneNumberInputTextXVM: InputTextXVM;
    emailInputTextXVM: InputTextXVM;
    endCustomerInputTextXVM: InputTextXVM;
    submitButtonXVM: ButtonXVM;
}
