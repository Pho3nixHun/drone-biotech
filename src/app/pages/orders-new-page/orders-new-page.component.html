@if (vm(); as vm) {
    <app-page-layout>
        <app-frame [vm]="vm" *transloco="let t">
            <form class="sm:p-8 sm:shadow-md" [formGroup]="formGroup">
                <!-- Internal order number input start -->
                @let hasInternalOrderNumberRequiredError =
                    formGroup.controls.internalOrderNumber.dirty &&
                    formGroup.controls.internalOrderNumber.hasError('required');
                <app-input-text
                    [vm]="vm.internalOrderNumberInputTextXVM"
                    [invalid]="hasInternalOrderNumberRequiredError"
                    [formControl]="formGroup.controls.internalOrderNumber"
                >
                    @if (hasInternalOrderNumberRequiredError) {
                        <span leftAssistiveText>{{
                            t(vm.requiredAssistiveTextKey)
                        }}</span>
                    }
                </app-input-text>
                <!-- Internal order number input end -->

                <div [formGroup]="formGroup.controls.contact">
                    <!-- Name input start -->
                    @let hasNameRequiredError =
                        formGroup.controls.contact.controls.name.dirty &&
                        formGroup.controls.contact.controls.name.hasError(
                            'required'
                        );
                    <app-input-text
                        [vm]="vm.nameInputTextXVM"
                        [invalid]="hasNameRequiredError"
                        [formControl]="formGroup.controls.contact.controls.name"
                    >
                        @if (hasNameRequiredError) {
                            <span leftAssistiveText>{{
                                t(vm.requiredAssistiveTextKey)
                            }}</span>
                        }
                    </app-input-text>
                    <!-- Name input end -->

                    <!-- Phone number input start -->
                    @let hasPhoneNumberRequiredError =
                        formGroup.controls.contact.controls.phoneNumber.dirty &&
                        formGroup.controls.contact.controls.phoneNumber.hasError(
                            'required'
                        );
                    <app-input-text
                        [vm]="vm.phoneNumberInputTextXVM"
                        [invalid]="hasPhoneNumberRequiredError"
                        [formControl]="
                            formGroup.controls.contact.controls.phoneNumber
                        "
                    >
                        @if (hasPhoneNumberRequiredError) {
                            <span leftAssistiveText>{{
                                t(vm.requiredAssistiveTextKey)
                            }}</span>
                        }
                    </app-input-text>
                    <!-- Phone number input end -->

                    <!-- Email input start -->
                    @let hasEmailRequiredError =
                        formGroup.controls.contact.controls.email.dirty &&
                        formGroup.controls.contact.controls.email.hasError(
                            'required'
                        );
                    <app-input-text
                        [vm]="vm.emailInputTextXVM"
                        [invalid]="hasEmailRequiredError"
                        [formControl]="
                            formGroup.controls.contact.controls.email
                        "
                    >
                        @if (hasEmailRequiredError) {
                            <span leftAssistiveText>{{
                                t(vm.requiredAssistiveTextKey)
                            }}</span>
                        }
                    </app-input-text>
                    <!-- Email input end -->
                </div>

                <!-- End customer input start-->
                @let hasEndCustomerRequiredError =
                    formGroup.controls.endCustomer.dirty &&
                    formGroup.controls.endCustomer.hasError('required');
                <app-input-text
                    [vm]="vm.endCustomerInputTextXVM"
                    [invalid]="hasEndCustomerRequiredError"
                    [formControl]="formGroup.controls.endCustomer"
                >
                    @if (hasEndCustomerRequiredError) {
                        <span leftAssistiveText>{{
                            t(vm.requiredAssistiveTextKey)
                        }}</span>
                    }
                </app-input-text>
                <!-- End customer input end-->

                <!-- Missions form control start -->
                @let hasAreasDataMinArrayLengthError =
                    formGroup.controls.areasData.touched &&
                    formGroup.controls.areasData.hasError('minArrayLength');

                <app-areas-data-form-control
                    #areasDataFormControl
                    [vm]="vm.areasDataFormControlVM"
                    [invalid]="hasAreasDataMinArrayLengthError"
                    [formControl]="formGroup.controls.areasData"
                />

                <!-- Missions form control end -->

                <app-button
                    [vm]="vm.submitButtonXVM"
                    [disabled]="formGroup.invalid"
                    (click)="submitForm()"
                >
                    @if (vm.submitButtonXVM.textKey; as textKey) {
                        {{ t(textKey) }}
                    }
                </app-button>
            </form>
        </app-frame>
    </app-page-layout>
}
