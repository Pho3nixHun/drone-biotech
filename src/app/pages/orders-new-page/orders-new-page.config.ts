import { Provider } from '@angular/core';
import {
    ENTRY_POINT_MARKER_OPTIONS,
    INFO_WINDOW_OPTIONS,
    MAP_OPTIONS,
    POLYGON_OPTIONS,
} from '@providers/google-maps-provider';

export const ORDERS_NEW_PAGE_PROVIDERS: Provider[] = [
    {
        provide: MAP_OPTIONS,
        useValue: {
            mapTypeId: 'roadmap',
            disableDoubleClickZoom: true,
            clickableIcons: false,
            mapId: 'DEMO_MAP_ID',
            zoom: 14,
            streetViewControl: false,
            mapTypeControlOptions: {
                position: google.maps.ControlPosition.TOP_RIGHT,
            },
        },
    },
    {
        provide: POLYGON_OPTIONS,
        useValue: {
            clickable: false,
            draggable: false,
            editable: true,
            fillOpacity: 0.5,
            strokeColor: 'blue',
            fillColor: 'blue',
        },
    },
    {
        provide: ENTRY_POINT_MARKER_OPTIONS,
        useValue: {
            gmpDraggable: true,
        },
    },

    {
        provide: INFO_WINDOW_OPTIONS,
        useValue: {
            disableAutoPan: true,
            headerDisabled: true,
        },
    },
];
