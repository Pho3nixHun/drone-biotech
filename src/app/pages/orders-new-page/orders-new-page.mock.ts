import { MatIcon } from '@interfaces/mat-icon.enum';
import { OrdersNewPageVM } from './orders-new-page-vm.model';

export const ordersNewPageVMDefault: OrdersNewPageVM = {
    frameXVM: {
        titleKey: 'OrdersNewPage.title',
        submitButtonXVM: {
            type: 'withText',
            textKey: 'OrdersNewPage.submitButtonText',
            secondary: false,
            variant: 'fill',
        },
        emailLabelKey: 'OrdersNewPage.emailLabel',
        endCustomerLabelKey: 'OrdersNewPage.endCustomerLabel',
        nameLabelKey: 'OrdersNewPage.nameLabel',
        phoneNumberLabelKey: 'OrdersNewPage.phoneNumberLabel',
        internalOrderNumberLabelKey: 'OrdersNewPage.internalOrderNumberLabel',
        areasDataLabelKey: 'OrdersNewPage.areasDataLabel',

        areasDataFormControlVM: {
            addButtonVM: {
                type: 'withIcon',
                icon: MatIcon.ADD,
                variant: 'ghost',
            },
            deleteButtonVM: {
                type: 'withIcon',
                icon: MatIcon.DELETE,
                secondary: true,
                variant: 'ghost',
            },
            editButtonVM: {
                type: 'withIcon',
                icon: MatIcon.EDIT,
                secondary: true,
                variant: 'ghost',
            },
            addAreaDataDialogVM: {
                type: 'areaDataDialogVM',
                closeButtonVM: {
                    type: 'withIcon',
                    icon: MatIcon.CLOSE,
                    secondary: true,
                    variant: 'ghost',
                },
                cancelButtonVM: {
                    type: 'withText',
                    textKey: 'OrdersNewPage.areasDataDialog.cancelButtonText',
                    secondary: true,
                    variant: 'ghost',
                },
                coordinatesLabelKey:
                    'OrdersNewPage.areasDataDialog.coordinatesLabel',
                commentLabelKey: 'OrdersNewPage.areasDataDialog.commentLabel',
                missionNameLabelKey:
                    'OrdersNewPage.areasDataDialog.missionNameLabel',
                submitButtonVM: {
                    type: 'withText',
                    textKey: 'OrdersNewPage.areasDataDialog.submitButtonText',
                    variant: 'fill',
                },
                applicationDateLabelKey:
                    'OrdersNewPage.areasDataDialog.applicationDateLabel',
                dosePerHqLabelKey:
                    'OrdersNewPage.areasDataDialog.dosePerHqLabel',
                entryPointLabelKey:
                    'OrdersNewPage.areasDataDialog.entryPointLabelKey',
                targetAreaLabelKey:
                    'OrdersNewPage.areasDataDialog.targetAreaLabelKey',
                mapFormControlVM: {
                    defaultCenter: null,
                    mapSearchInputFormControlVM: {
                        distanceValueKey:
                            'OrdersNewPage.areasDataDialog.mapSearchInputFormControlVM.distanceValue',
                        placeholderKey:
                            'OrdersNewPage.areasDataDialog.mapSearchInputFormControlVM.placeholder',
                    },
                    mapAreaSelectFormControlVM: {
                        areaValueKey:
                            'OrdersNewPage.areasDataDialog.mapAreaSelectFormControl.areaValueKey',
                        addButtonVM: {
                            type: 'withText',
                            variant: 'ghost',
                            textKey:
                                'OrdersNewPage.areasDataDialog.mapAreaSelectFormControl.addButtonText',
                        },
                        editButtonVM: {
                            type: 'withText',
                            secondary: true,
                            variant: 'ghost',
                            textKey:
                                'OrdersNewPage.areasDataDialog.mapAreaSelectFormControl.editButtonText',
                        },
                        deleteButtonVM: {
                            type: 'withText',
                            secondary: true,
                            variant: 'ghost',
                            textKey:
                                'OrdersNewPage.areasDataDialog.mapAreaSelectFormControl.deleteButtonText',
                        },
                        coordinatesLabelKey:
                            'OrdersNewPage.areasDataDialog.mapAreaSelectFormControl.coordinatesLabel',
                    },
                    mapPointSelectFormControlVM: {
                        addButtonVM: {
                            type: 'withText',
                            variant: 'ghost',
                            textKey:
                                'OrdersNewPage.areasDataDialog.mapPointSelectFormControl.addButtonText',
                        },
                        deleteButtonVM: {
                            type: 'withText',
                            secondary: true,
                            variant: 'ghost',
                            textKey:
                                'OrdersNewPage.areasDataDialog.mapPointSelectFormControl.deleteButtonText',
                        },
                    },
                },
                areaData: null,
                titleKey: 'OrdersNewPage.areasDataDialog.title.add',
            },

            editAreaDataDialogVM: {
                type: 'areaDataDialogVM',
                closeButtonVM: {
                    type: 'withIcon',
                    icon: MatIcon.CLOSE,
                    secondary: true,
                    variant: 'ghost',
                },
                coordinatesLabelKey:
                    'OrdersNewPage.areasDataDialog.coordinatesLabel',
                commentLabelKey: 'OrdersNewPage.areasDataDialog.commentLabel',
                missionNameLabelKey:
                    'OrdersNewPage.areasDataDialog.missionNameLabel',
                submitButtonVM: {
                    type: 'withText',
                    textKey: 'OrdersNewPage.areasDataDialog.submitButtonText',
                    variant: 'fill',
                },
                cancelButtonVM: {
                    type: 'withText',
                    textKey: 'OrdersNewPage.areasDataDialog.cancelButtonText',
                    secondary: true,
                    variant: 'ghost',
                },
                applicationDateLabelKey:
                    'OrdersNewPage.areasDataDialog.applicationDateLabel',
                dosePerHqLabelKey:
                    'OrdersNewPage.areasDataDialog.dosePerHqLabel',
                entryPointLabelKey:
                    'OrdersNewPage.areasDataDialog.entryPointLabelKey',
                targetAreaLabelKey:
                    'OrdersNewPage.areasDataDialog.targetAreaLabelKey',
                mapFormControlVM: {
                    defaultCenter: null,
                    mapSearchInputFormControlVM: {
                        distanceValueKey:
                            'OrdersNewPage.areasDataDialog.mapSearchInputFormControlVM.distanceValue',
                        placeholderKey:
                            'OrdersNewPage.areasDataDialog.mapSearchInputFormControlVM.placeholder',
                    },
                    mapAreaSelectFormControlVM: {
                        areaValueKey:
                            'OrdersNewPage.areasDataDialog.mapAreaSelectFormControl.areaValueKey',
                        addButtonVM: {
                            type: 'withText',
                            variant: 'ghost',
                            textKey:
                                'OrdersNewPage.areasDataDialog.mapAreaSelectFormControl.addButtonText',
                        },
                        editButtonVM: {
                            type: 'withText',
                            secondary: true,
                            variant: 'ghost',
                            textKey:
                                'OrdersNewPage.areasDataDialog.mapAreaSelectFormControl.editButtonText',
                        },
                        deleteButtonVM: {
                            type: 'withText',
                            secondary: true,
                            variant: 'ghost',
                            textKey:
                                'OrdersNewPage.areasDataDialog.mapAreaSelectFormControl.deleteButtonText',
                        },
                        coordinatesLabelKey:
                            'OrdersNewPage.areasDataDialog.mapAreaSelectFormControl.coordinatesLabel',
                    },
                    mapPointSelectFormControlVM: {
                        addButtonVM: {
                            type: 'withText',
                            variant: 'ghost',
                            textKey:
                                'OrdersNewPage.areasDataDialog.mapPointSelectFormControl.addButtonText',
                        },
                        deleteButtonVM: {
                            type: 'withText',
                            secondary: true,
                            variant: 'ghost',
                            textKey:
                                'OrdersNewPage.areasDataDialog.mapPointSelectFormControl.deleteButtonText',
                        },
                    },
                },

                areaData: null,
                titleKey: 'OrdersNewPage.areasDataDialog.title.edit',
            },

            deleteDialogVM: {
                type: 'deleteDialogVM',
                closeButtonVM: {
                    type: 'withIcon',
                    icon: MatIcon.CLOSE,
                    secondary: true,
                    variant: 'ghost',
                },
                titleKey: 'OrdersNewPage.deleteDialog.title',
                warningTextKey: 'OrdersNewPage.deleteDialog.warningText',
                deleteButtonVM: {
                    type: 'withText',
                    textKey: 'OrdersNewPage.deleteDialog.submitButtonText',
                    variant: 'fill',
                },
                cancelButtonVM: {
                    type: 'withText',
                    textKey: 'OrdersNewPage.deleteDialog.cancelButtonText',
                    secondary: true,
                    variant: 'ghost',
                },
            },
            actionsHeaderKey:
                'OrdersNewPage.areasDataFormControl.actionsHeader',
            targetAreaSizeHeaderKey:
                'OrdersNewPage.areasDataFormControl.targetAreaSizeHeader',
            entryPointHeaderKey:
                'OrdersNewPage.areasDataFormControl.entryPointHeader',
            doseHeaderKey: 'OrdersNewPage.areasDataFormControl.doseHeader',
            applicationDateHeaderKey:
                'OrdersNewPage.areasDataFormControl.applicationDateHeader',
            trichogrammaRequirementHeaderKey:
                'OrdersNewPage.areasDataFormControl.trichogrammaRequirementHeader',
            distanceFromHeadOfficeHeaderKey:
                'OrdersNewPage.areasDataFormControl.distanceFromHeadOfficeHeader',
            totalDistanceFromHeadOfficeHeaderKey:
                'OrdersNewPage.areasDataFormControl.totalDistanceFromHeadOfficeHeader',
            totalTargetAreaSizeHeaderKey:
                'OrdersNewPage.areasDataFormControl.totalTargetAreaSizeHeader',
            totalTrichogrammaRequirementHeaderKey:
                'OrdersNewPage.areasDataFormControl.totalTrichogrammaRequirementHeader',
            applicationDateValueKey:
                'OrdersNewPage.areasDataFormControl.applicationDateValue',
            distanceFromHeadOfficeValueKey:
                'OrdersNewPage.areasDataFormControl.distanceFromHeadOfficeValue',
            doseValueKey: 'OrdersNewPage.areasDataFormControl.doseValue',
            entryPointValueKey: '',
            targetAreaSizeValueKey:
                'OrdersNewPage.areasDataFormControl.targetAreaSizeValue',
            trichogrammaRequirementValueKey:
                'OrdersNewPage.areasDataFormControl.trichogrammaRequirementValue',
            commentHeaderKey:
                'OrdersNewPage.areasDataFormControl.commentHeader',
            missionHeaderKey:
                'OrdersNewPage.areasDataFormControl.missionHeader',
        },
    },
};

export const enMock = {
    closeButtonText: 'close',
    coordinatesLabel: 'coords',
    commentLabel: 'comment',
    missionLabel: 'mission',
    frameTitle: 'tit',
    submitButtonText: 'sub',
    emailLabel: 'email',
    endCustomerLabel: 'endCustomer',
    nameLabel: 'nameLab',
    phoneNumberLabel: 'phoneNumber',
    internalOrderNumberLabel: 'internal',
    areasDataLabel: 'areasData',
    cancelButtonText: 'cancelButton',
    applicationDateLabel: 'applicationDate',
    dosePerHqLabel: 'dosePer',
    entryPointLabel: 'entryPoint',
    targetAreaLabel: 'targetArea',
    addButtonText: 'addButton',
    deleteButtonText: 'deleteButton',
    title: 'ti',
    type: 'areaDataDialogVM',
    actionsHeader: 'actions',
    targetAreaSizeHeader: 'targetAreaSize',
    entryPointHeader: 'entryPoint',
    doseHeader: 'dose',
    applicationDateHeader: 'application',
    trichogrammaRequirementHeader: 'trichogrammaRequirement',
    distanceFromHeadOfficeHeader: 'distanceFromHeadOffice',
    totalDistanceFromHeadOfficeHeader: 'totalDistanceFromHeadOffice',
    totalTargetAreaSizeHeader: 'totalTargetAreaSize',
    totalTrichogrammaRequirementHeader: 'totalTrichogrammaRequirement',
    applicationDateValue: 'applicationDate',
    distanceFromHeadOfficeValue: 'distanceFromHeadOffice',
    doseValue: 'doseValue',
    entryPointValue: '',
    targetAreaSizeValue: 'targetAreaSize',
    trichogrammaRequirementValue: 'trichogrammaRequirement',
    warning: 'warn',
    placeholder: 'place',
    distanceValue: 'distance',
    commentHeader: 'comment',
    missionHeader: 'mission',
    areaValue: 'area',
    editButton: 'edit',
};

export const ordersNewPageVMMock: OrdersNewPageVM = {
    frameXVM: {
        titleKey: enMock.frameTitle,
        submitButtonXVM: {
            type: 'withText',
            variant: 'fill',
            secondary: false,
            textKey: enMock.submitButtonText,
        },
        emailLabelKey: enMock.emailLabel,
        endCustomerLabelKey: enMock.endCustomerLabel,
        nameLabelKey: enMock.nameLabel,
        phoneNumberLabelKey: enMock.phoneNumberLabel,
        internalOrderNumberLabelKey: enMock.internalOrderNumberLabel,
        areasDataLabelKey: enMock.areasDataLabel,

        areasDataFormControlVM: {
            addButtonVM: {
                type: 'withIcon',
                secondary: false,
                variant: 'ghost',
                icon: MatIcon.ADD,
            },
            editButtonVM: {
                type: 'withIcon',
                secondary: false,
                variant: 'ghost',
                icon: MatIcon.EDIT,
            },
            deleteButtonVM: {
                type: 'withIcon',
                secondary: false,
                variant: 'ghost',
                icon: MatIcon.DELETE,
            },

            addAreaDataDialogVM: {
                type: 'areaDataDialogVM',
                closeButtonVM: {
                    type: 'withText',
                    textKey: enMock.closeButtonText,
                    secondary: false,
                    variant: 'fill',
                },
                coordinatesLabelKey: enMock.coordinatesLabel,
                commentLabelKey: enMock.commentLabel,
                missionNameLabelKey: enMock.missionLabel,
                cancelButtonVM: {
                    type: 'withText',
                    textKey: enMock.cancelButtonText,
                    secondary: false,
                    variant: 'fill',
                },
                submitButtonVM: {
                    type: 'withText',
                    textKey: enMock.cancelButtonText,
                    secondary: false,
                    variant: 'fill',
                },
                applicationDateLabelKey: enMock.applicationDateLabel,
                dosePerHqLabelKey: enMock.dosePerHqLabel,
                entryPointLabelKey: enMock.entryPointLabel,
                targetAreaLabelKey: enMock.targetAreaLabel,
                mapFormControlVM: {
                    defaultCenter: null,
                    mapSearchInputFormControlVM: {
                        placeholderKey: enMock.placeholder,
                        distanceValueKey: enMock.distanceValue,
                    },
                    mapAreaSelectFormControlVM: {
                        areaValueKey: enMock.areaValue,
                        addButtonVM: {
                            type: 'withText',
                            secondary: true,
                            variant: 'ghost',
                            textKey: enMock.addButtonText,
                        },
                        editButtonVM: {
                            type: 'withText',
                            secondary: true,
                            variant: 'ghost',
                            textKey: enMock.deleteButtonText,
                        },
                        deleteButtonVM: {
                            type: 'withText',
                            secondary: true,
                            variant: 'ghost',
                            textKey: enMock.editButton,
                        },
                        coordinatesLabelKey: enMock.coordinatesLabel,
                    },
                    mapPointSelectFormControlVM: {
                        addButtonVM: {
                            type: 'withText',
                            secondary: true,
                            variant: 'ghost',
                            textKey: enMock.addButtonText,
                        },
                        deleteButtonVM: {
                            type: 'withText',
                            secondary: true,
                            variant: 'ghost',
                            textKey: enMock.deleteButtonText,
                        },
                    },
                },
                areaData: null,
                titleKey: enMock.title,
            },

            editAreaDataDialogVM: {
                type: 'areaDataDialogVM',
                closeButtonVM: {
                    type: 'withIcon',
                    icon: MatIcon.CLOSE,
                    secondary: false,
                    variant: 'fill',
                },
                coordinatesLabelKey: enMock.coordinatesLabel,
                commentLabelKey: enMock.commentLabel,
                missionNameLabelKey: enMock.missionLabel,
                cancelButtonVM: {
                    type: 'withText',
                    textKey: enMock.cancelButtonText,
                    secondary: false,
                    variant: 'fill',
                },
                submitButtonVM: {
                    type: 'withText',
                    textKey: enMock.submitButtonText,
                    secondary: false,
                    variant: 'fill',
                },
                applicationDateLabelKey: enMock.applicationDateLabel,
                dosePerHqLabelKey: enMock.dosePerHqLabel,
                entryPointLabelKey: enMock.entryPointLabel,
                targetAreaLabelKey: enMock.targetAreaLabel,
                mapFormControlVM: {
                    defaultCenter: null,

                    mapSearchInputFormControlVM: {
                        placeholderKey: enMock.placeholder,
                        distanceValueKey: enMock.distanceValue,
                    },
                    mapAreaSelectFormControlVM: {
                        areaValueKey: enMock.areaValue,
                        addButtonVM: {
                            type: 'withText',
                            secondary: true,
                            variant: 'ghost',
                            textKey: enMock.addButtonText,
                        },
                        editButtonVM: {
                            type: 'withText',
                            secondary: true,
                            variant: 'ghost',
                            textKey: enMock.deleteButtonText,
                        },
                        deleteButtonVM: {
                            type: 'withText',
                            secondary: true,
                            variant: 'ghost',
                            textKey: enMock.editButton,
                        },
                        coordinatesLabelKey: enMock.coordinatesLabel,
                    },
                    mapPointSelectFormControlVM: {
                        addButtonVM: {
                            type: 'withText',
                            secondary: true,
                            variant: 'ghost',
                            textKey: enMock.addButtonText,
                        },
                        deleteButtonVM: {
                            type: 'withText',
                            secondary: true,
                            variant: 'ghost',
                            textKey: enMock.deleteButtonText,
                        },
                    },
                },

                areaData: null,
                titleKey: enMock.title,
            },

            deleteDialogVM: {
                type: 'deleteDialogVM',
                closeButtonVM: {
                    type: 'withIcon',
                    icon: MatIcon.CLOSE,
                    secondary: true,
                    variant: 'ghost',
                },
                titleKey: enMock.title,
                warningTextKey: enMock.warning,
                cancelButtonVM: {
                    type: 'withText',
                    textKey: enMock.cancelButtonText,
                    secondary: false,
                    variant: 'fill',
                },
                deleteButtonVM: {
                    type: 'withText',
                    textKey: enMock.submitButtonText,
                    secondary: false,
                    variant: 'fill',
                },
            },
            actionsHeaderKey: enMock.actionsHeader,
            targetAreaSizeHeaderKey: enMock.targetAreaSizeHeader,
            entryPointHeaderKey: enMock.entryPointHeader,
            doseHeaderKey: enMock.doseHeader,
            applicationDateHeaderKey: enMock.applicationDateHeader,
            trichogrammaRequirementHeaderKey:
                enMock.trichogrammaRequirementHeader,
            distanceFromHeadOfficeHeaderKey:
                enMock.distanceFromHeadOfficeHeader,
            totalDistanceFromHeadOfficeHeaderKey:
                enMock.totalDistanceFromHeadOfficeHeader,
            totalTargetAreaSizeHeaderKey: enMock.totalTargetAreaSizeHeader,
            totalTrichogrammaRequirementHeaderKey:
                enMock.totalTrichogrammaRequirementHeader,
            applicationDateValueKey: enMock.applicationDateHeader,
            distanceFromHeadOfficeValueKey: enMock.distanceFromHeadOfficeHeader,
            doseValueKey: enMock.doseValue,
            entryPointValueKey: enMock.entryPointValue,
            targetAreaSizeValueKey: enMock.targetAreaSizeValue,
            trichogrammaRequirementValueKey:
                enMock.trichogrammaRequirementValue,
            commentHeaderKey: enMock.commentHeader,
            missionHeaderKey: enMock.missionHeader,
        },
    },
};
