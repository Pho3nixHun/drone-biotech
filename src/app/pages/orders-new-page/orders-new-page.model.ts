import { AreasDataFormControlVM } from './components/areas-data-form-control/areas-data-form-control.model';
import { FrameVM } from '@components/frame/frame.model';

export interface OrdersNewPageVM {
    frameXVM: FrameXVM;
}

export interface OrdersNewFrame extends FrameVM {
    areasDataFormControlVM: AreasDataFormControlVM;
    submitButtonTextKey: string;
    internalOrderNumberLabelKey: string;
    nameLabelKey: string;
    phoneNumberLabelKey: string;
    emailLabelKey: string;
    endCustomerLabelKey: string;
}

type FrameXVM = OrdersNewFrame;
