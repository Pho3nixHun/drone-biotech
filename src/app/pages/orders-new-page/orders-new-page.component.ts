import { NgClass } from '@angular/common';
import { Component, inject } from '@angular/core';
import { FormBuilder, ReactiveFormsModule, Validators } from '@angular/forms';
import { FrameComponent } from '@components/frame/frame.component';
import { OrdersNewPageService } from './orders-new-page.service';
import { LocationStoreModule } from 'src/app/stores/location/location.module';
import { TranslocoModule } from '@jsverse/transloco';
import { AreasDataFormControlComponent } from './components/areas-data-form-control/areas-data-form-control.component';
import { AreaData } from './components/areas-data-form-control/components/area-data-dialog/area-data-dialog.model';
import { PageLayoutComponent } from '@components/page-layout/page-layout.component';
import { ORDERS_NEW_PAGE_PROVIDERS } from './orders-new-page.config';

/**
 * OrdersNewPageComponent
 *
 * Type: Container
 *
 * Scope:
 * - Renders a page that contains a complex form to handle new orders set by the customer.
 * - Passes relevant data, like config, texts.
 *
 * Out-of-Scope:
 * - Does not handle the internal logic or styling of other components.
 * - Not responsible for the detailed presentation logic.
 *
 * Purpose (optional):
 * - To serve as a smart container component that integrates business logic, including data fetching and presentation, to create a cohesive user interface.
 */

@Component({
    selector: 'app-orders-new-page',
    imports: [
        NgClass,
        ReactiveFormsModule,
        TranslocoModule,
        LocationStoreModule,
        FrameComponent,
        AreasDataFormControlComponent,
        PageLayoutComponent,
    ],
    templateUrl: './orders-new-page.component.html',
    providers: [ORDERS_NEW_PAGE_PROVIDERS],
})
export class OrdersNewPageComponent {
    private readonly ordersService = inject(OrdersNewPageService);
    private readonly fb = inject(FormBuilder);
    protected readonly vm = this.ordersService.getVM();

    public areasDataFormGroup = this.fb.group({
        internalOrderNumber: this.fb.control<string>('', Validators.required),
        contact: this.fb.group({
            name: this.fb.control<string>('', Validators.required),
            phoneNumber: this.fb.control<string>('', [Validators.required]),
            email: this.fb.control<string>('', [
                Validators.required,
                Validators.email,
            ]),
        }),
        endCustomer: this.fb.control<string>('', Validators.required),
        areasData: this.fb.control<AreaData[]>([], [Validators.required]),
    });

    protected submitForm() {
        if (this.areasDataFormGroup.valid) {
            this.areasDataFormGroup.reset();
        }
    }
}
