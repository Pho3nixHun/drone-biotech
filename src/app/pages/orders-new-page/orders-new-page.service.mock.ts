import { Injectable, Signal, signal } from '@angular/core';
import { OrdersNewPageVM } from './orders-new-page-vm.model';
import { OrdersNewPageService } from './orders-new-page.service';

@Injectable({
    providedIn: 'root',
})
export class OrdersNewPageMockService {
    public getVM(): Signal<OrdersNewPageVM | undefined> {
        return vmSignal;
    }
}

export const updateVMSignal = (vm: OrdersNewPageVM | undefined) => {
    vmSignal.set(vm);
};

const vmSignal = signal<OrdersNewPageVM | undefined>(undefined);

export const provideOrdersNewPageMock = () => ({
    provide: OrdersNewPageService,
    useClass: OrdersNewPageMockService,
});
