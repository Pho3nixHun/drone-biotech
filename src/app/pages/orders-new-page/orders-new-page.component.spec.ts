import { ComponentFixture, TestBed } from '@angular/core/testing';
import { getTranslocoModule } from 'transloco-testing.module';
import { enMock, ordersNewPageVMMock } from './orders-new-page.mock';
import { OrdersNewPageComponent } from './orders-new-page.component';
import {
    provideOrdersNewPageMockService,
    updateVMSignal,
} from './orders-new-page.service.mock';
import { LocationStoreMockModule } from 'src/app/stores/location/location.module';

describe('OrdersNewPageComponent', () => {
    let component: OrdersNewPageComponent;
    let fixture: ComponentFixture<OrdersNewPageComponent>;
    let compiled: HTMLElement;

    beforeEach(async () => {
        await TestBed.configureTestingModule({
            imports: [
                OrdersNewPageComponent,
                LocationStoreMockModule,
                getTranslocoModule({
                    langs: { en: enMock },
                    translocoConfig: {
                        availableLangs: ['en'],
                        defaultLang: 'en',
                    },
                }),
            ],
            providers: [provideOrdersNewPageMockService()],
        }).compileComponents();

        fixture = TestBed.createComponent(OrdersNewPageComponent);
        compiled = fixture.debugElement.nativeElement;
        component = fixture.componentInstance;
    });

    //Snapshot testing
    it('should not render the template when the VM is not provided', () => {
        //Arrange
        updateVMSignal(undefined);
        //Act
        fixture.detectChanges();
        //Assert
        expect(compiled).toMatchSnapshot();
    });

    //Snapshot testing
    it('should render the template when the VM is provided', () => {
        //Arrange
        updateVMSignal(ordersNewPageVMMock);

        //Act
        fixture.detectChanges();

        //Assert
        expect(compiled).toMatchSnapshot();
    });

    it('should set the areasDataFormGroup to invalid if there is no value in the form', () => {
        // Arrange

        // Act

        // Assert
        expect(component.areasDataFormGroup.touched).toBe(false);
        expect(component.areasDataFormGroup.invalid).toBe(true);
    });

    it('should set the areasDataFormGroup to valid if the form is fulfilled with correct values', () => {
        //Arrange

        //Act
        component.areasDataFormGroup.setValue({
            areasData: [
                {
                    applicationDate: new Date(),
                    dosePerHq: 10,
                    entryPoint: { lat: 10, lng: 10 },
                    id: 10,
                    targetArea: [
                        { lat: 10, lng: 10 },
                        { lat: 20, lng: 20 },
                        { lat: 30, lng: 30 },
                    ],
                },
            ],
            contact: {
                email: 'test@gmail.com',
                name: 'Joe',
                phoneNumber: '06301111111',
            },
            endCustomer: 'customer',
            internalOrderNumber: 'number',
        });

        // Assert
        expect(component.areasDataFormGroup.valid).toBe(true);
    });
});
