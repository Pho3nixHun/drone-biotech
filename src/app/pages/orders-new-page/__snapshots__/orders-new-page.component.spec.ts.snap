// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OrdersNewPageComponent should not render the template when the VM is not provided 1`] = `
<div
  id="root0"
  ng-version="18.2.8"
>
  <!--container-->
</div>
`;

exports[`OrdersNewPageComponent should render the template when the VM is provided 1`] = `
<div
  id="root1"
  ng-version="18.2.8"
>
  <app-page-section>
    <section
      class="mt-72 md:mt-25"
    >
      <!--container-->
    </section>
  </app-page-section>
  <!--container-->
</div>
`;
